### PR TITLE
Fix: Onyx.update order

### PR DESCRIPTION
**Pull request by chrispader**
@neil-marcellini 

### Details

Fixes a problem where `Onyx.update` seems to not apply updates in order. This was due to the merge queue not being cleared when we call `Onyx.set`.

Additionally, we returned early from `Onyx.set` and `Onyx.merge`, when we removed keys from storage and therefore didn't "broadcast" the update and update subscribers.

This PR fixed the problem with the order of `Onyx.update` by clearing the merge queue on every `Onyx.set` and additionally update subscribers with the cleared (null) value after removing a key intentionally with `Onyx.set(SOME_KEY, null)` or `Onyx.merge(SOME_KEY, null)`.

This PR also fixes some inconsistencies with key removal in `Onyx.multiSet` and `Onyx.multiMerge`.

### Related Issues

https://github.com/Expensify/react-native-onyx/issues/266

### Automated Tests

Added a Test "should apply updates in order with Onyx.update" in `onyxTest`, which tests `Onyx.update` for applying the changes in the correct order.

### Manual Tests

1. Open NewDot
2. Turn off network
3. Create a workspace
4. Delete the workspace
5. Turn on network
6. Check that the workspace is removed and disappears and it doesn't flash

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/20173411/9bdf48ad-86a9-4d39-bdb8-aea495c2128c

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/20173411/5c174f7d-84ac-4317-ba49-b3c49c07966b

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

Added screen recordings for web and iOS web

### CODE REVIEW

**Comment by chrispader** [comment]
@neil-marcellini this PR should fix the problems with `Onyx.update` order and the "replay" effect. Feel free to review this, i'm gonna add screenshots tmrw. Testing fine though 👍

File: lib/Onyx.js
Comment (neil-marcellini) [inline lib/Onyx.js:1062]: NAB: It seems pointless to have this `wasRemoved` flag when the caller could just check if the value is null to determine if the key was removed.

File: lib/Onyx.js
Comment (neil-marcellini) [inline lib/Onyx.js:1102]: If we removed keys by setting their values to null, don't we need to update storage with the new object without those keys?

File: lib/Onyx.js
Comment (neil-marcellini) [inline lib/Onyx.js:1035]: Why would we not update the cache when the key is removed?

File: lib/Onyx.js
Comment (neil-marcellini) [inline lib/Onyx.js:1117]: I don't see where we're returning the `wasRemoved` boolean.

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Looking really good, thank you! Would you please write manual test and link to the video showing it working for creating and deleting a workspace offline? That was great, so nice to see it finally working!

I do have a handful of comments and questions, and I would like to see some small changes to clean up a bit please.

File: lib/Onyx.js
Comment (jasperhuangg) [reply lib/Onyx.js:1035]: +1, I don't understand this

File: lib/Onyx.js
Comment (chrispader) [reply lib/Onyx.js:1102]: `wasRemoved` here only applies when the whole key was set to `null`, not when nested `null` values got removed.

The storage is already being updated, when the `remove` function is called here: https://github.com/margelo/react-native-onyx/blob/ac8faed9890bca04f889f93e0f249191ee6301b5/lib/Onyx.js#L1061

That's why we have to omit calling `Storage.setItem` here again

File: lib/Onyx.js
Comment (chrispader) [reply lib/Onyx.js:1035]: Same as in the the other thread, the call to `remove` already drops the key from cache: https://github.com/margelo/react-native-onyx/blob/ac8faed9890bca04f889f93e0f249191ee6301b5/lib/Onyx.js#L965

That's why we have to omit updating the storage here, and instead just add the key to the list of accessed keys.

File: lib/Onyx.js
Comment (chrispader) [reply lib/Onyx.js:1062]: The reason i introduced this extra flag was to have a consistent variable/condition of whether the key got removed as part of an operation. We check for this in multiple spots (both of your other comments), and i felt that having a dedicated `wasRemoved` flag was better than always checking for `if (valueAfterRemoving === null)`.

File: lib/Onyx.js
Comment (chrispader) [reply lib/Onyx.js:1117]: updated the comment! this was a relict from previous changes

**Comment by chrispader** [comment]
@neil-marcellini the video showcasing the fix of the replay effect is under "Screenshots"

**Comment by jasperhuangg** [comment]
Heading OOO for the holidays, so it might be best to assign another engineer if you want another review

File: lib/Onyx.js
Comment (neil-marcellini) [reply lib/Onyx.js:1062]: I generally prefer limiting the number of parameters. Also, wasRemoved is always simply `_.isNull(value)`, except in one case where we pass the batched changes. In fact you have a comment there explaining that we're checking if the batched changes are null. That's kind of silly.

File: lib/Onyx.js
Comment (neil-marcellini) [reply lib/Onyx.js:1102]: Ooooooh I see. Thank you for explaining.

File: lib/Onyx.js
Comment (neil-marcellini) [reply lib/Onyx.js:1035]: Makes sense thanks

**Review by neil-marcellini — APPROVED** [review]
Thanks for addressing my concerns, it makes sense now and looks good. This is awesome, and I think it's going to fix a lot of rare bugs we had lurking in the App!

File: lib/Onyx.js
Comment (neil-marcellini) [reply lib/Onyx.js:1062]: I'll leave it up to you if you want to change this is a follow up PR.

**Pull request merged**
